src/EditorFeatures/Test/Snippets/RoslynLSPSnippetConvertTests.cs

```diff
@@ -0,0 +1,243 @@
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.RoslynLSPSnippetConverter)]
+        public Task TestExtendSnippetTextChangeForwardsForPlaceholderThenCaret()
+        {
+            var markup =
+@"[|if (true)
+{
+}|] {|placeholder:test|} $$";
+
+            var expectedLSPSnippet =
+@"if (true)
+{
+} ${1:test} $0";
+
+            return TestAsync(markup, expectedLSPSnippet);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.RoslynLSPSnippetConverter)]
+        public Task TestExtendSnippetTextChangeForwardsForCaretThenPlaceholder()
+        {
+            var markup =
+@"[|if (true)
+{
+}|] $$ {|placeholder:test|}";
+
+            var expectedLSPSnippet =
+@"if (true)
+{
+} $0 ${1:test}";
+
+            return TestAsync(markup, expectedLSPSnippet);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.RoslynLSPSnippetConverter)]
+        public Task TestExtendSnippetTextChangeBackwardsForPlaceholderThenCaret()
+        {
+            var markup =
+@"{|placeholder:test|} $$ [|if (true)
+{
+}|]";
+
+            var expectedLSPSnippet =
+@"${1:test} $0 if (true)
+{
+}";
+
+            return TestAsync(markup, expectedLSPSnippet);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.RoslynLSPSnippetConverter)]
+        public Task TestExtendSnippetTextChangeBackwardsForCaretThenPlaceholder()
+        {
+            var markup =
+@"$$ {|placeholder:test|} [|if (true)
+{
+}|]";
+
+            var expectedLSPSnippet =
+@"$0 ${1:test} if (true)
+{
+}";
+
+            return TestAsync(markup, expectedLSPSnippet);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.RoslynLSPSnippetConverter)]
+        public Task TestExtendSnippetTextChangeBackwardsForCaretForwardsForPlaceholder()
+        {
+            var markup =
+@"$$ [|if (true)
+{
+}|] {|placeholder:test|}";
+
+            var expectedLSPSnippet =
+@"$0 if (true)
+{
+} ${1:test}";
+
+            return TestAsync(markup, expectedLSPSnippet);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.RoslynLSPSnippetConverter)]
+        public Task TestExtendSnippetTextChangeBackwardsForPlaceholderForwardsForCaret()
+        {
+            var markup =
+@"{|placeholder:test|} [|if (true)
+{
+}|] $$";
+
+            var expectedLSPSnippet =
+@"${1:test} if (true)
+{
+} $0";
+
+            return TestAsync(markup, expectedLSPSnippet);
+        }
+
+        #endregion
+
+        #region LSP Snippet generation tests
+
+        [Fact, Trait(Traits.Feature, Traits.Features.RoslynLSPSnippetConverter)]
+        public Task TestForLoopSnippet()
+        {
+            var markup =
+@"[|for (var {|placeholder1:i|} = 0; {|placeholder1:i|} < {|placeholder2:length|}; {|placeholder1:i|}++)
+{$$
+}|]";
+
+            var expectedLSPSnippet =
+@"for (var ${1:i} = 0; ${1:i} < ${2:length}; ${1:i}++)
+{$0
+}";
+
+            return TestAsync(markup, expectedLSPSnippet);
+        }
+
+        #endregion
```

Comment:
nice.  these are great.

Reply:
it would be good to have some tests where tehre are more code before/after teh snippet.  consider a version of all these tests where the snippet is inside (and indented) in:

```c#
void M()
{
}
```

Should be easy to copy/paste to get the new tests, especially as now tehy're all declarative, so you don't need to update any test code beyond the string literal contents.